### PR TITLE
Adding feature of wireless provisioning and reprovisioning

### PR DESCRIPTION
**Pull request by abimishr**
**1. Git PR Link:** https://github.com/madhansansel/dnacenter-ansible/pull/263
**2. Problem Description:** Provisioning wireless devices to a site
 **3. Code Changes:**  Fixed the payload for the wireless provisioning API. Added conditions to check the re-provisioning part.
**4. Follow-up Tasks:** Will need to work on the assigning to site part for wireless devices.
**5.  Testing and Automation:** Tested for wireless WLC provisioning and re-provisioning.
TASK [Provision a wireless device to a site] ***********************************************************************************************************************************************************

PLAY RECAP *********************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

### CODE REVIEW

File: plugins/modules/provision_workflow_manager.py [hunk 1]
```diff
@@ -436,6 +485,52 @@ def get_execution_status(self, execution_id=None):
...
+            response = self.dnac_apply['exec'](
+                family="task",
+                function="get_business_api_execution_details",
+                params=params,
+                op_modifies=True
+            )
+            self.log("Response collected from 'get_business_api_execution_details' API is {0}".format(str(response)), "DEBUG")
+            self.log("Execution status for the execution id {0} is {1}".format(str(execution_id), str(response.get("status"))), "INFO")
+            if response.get('bapiError') or response.get("status") == "FAILURE":
+                if response.get("bapiError") == "Device was already provisioned , please use provision update API to reprovision the device":
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:516]: usually we won't leave space before comma..  Do we get bapiError like this? Can you check?

File: plugins/modules/provision_workflow_manager.py [hunk 2]
```diff
@@ -623,31 +718,38 @@ def get_wireless_params(self):
                 "managedAPLocations": self.validated_config[0].get("managed_ap_locations"),
             }
         ]
-        for ap_loc in wireless_params[0]["managedAPLocations"]:
+
+        if not wireless_params[0].get("managedAPLocations"):
+            msg = "Managed AP locations must be passed as a list of sites"
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:723]: Can we say, for example ...

File: plugins/modules/provision_workflow_manager.py [hunk 3]
```diff
@@ -683,6 +785,47 @@ def get_want(self):
...
+        try:
+            headers_payload = {"__persistbapioutput": "true"}
+            response = self.dnac_apply['exec'](
+                family="wireless",
+                function="provision_update",
+                op_modifies=True,
+                params={"payload": self.want.get("prov_params"),
+                        "headers": headers_payload}
+            )
+            self.log("Wireless provisioning response collected from 'provision' API is: {0}".format(str(response)), "DEBUG")
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:813]: from 'provision_update' API

File: plugins/modules/provision_workflow_manager.py [hunk 4]
```diff
@@ -683,6 +785,47 @@ def get_want(self):
...
+            execution_id = response.get("executionId")
+            provision_info = self.get_execution_status_wireless(execution_id=execution_id)
+            self.result["changed"] = True
+            self.result['msg'] = "Wireless device got re-provisioned successfully"
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:817]: Is it possible to add the name/ip of which wireless device has been successfully re-provisioned?

File: plugins/modules/provision_workflow_manager.py [hunk 5]
```diff
@@ -789,13 +932,28 @@ class instance for further use.
...
+                    family="wireless",
+                    function="provision",
+                    op_modifies=True,
+                    params={"payload": self.want.get("prov_params")}
+                )
+                self.log("Wireless provisioning response collected from 'provision' API is: {0}".format(str(response)), "DEBUG")
+                execution_id = response.get("executionId")
+                provision_info = self.get_execution_status_wireless(execution_id=execution_id)
+                self.result["changed"] = True
+                self.result['msg'] = "Wireless device got provisioned successfully"
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:946]: Is it possible to add the name/ip of which wireless device has been successfully provisioned?

File: plugins/modules/provision_workflow_manager.py [hunk 6]
```diff
@@ -789,13 +932,28 @@ class instance for further use.
...
+                self.result['diff'] = self.validated_config
+                self.result['response'] = execution_id
+                self.log(self.result['msg'], "INFO")
+                return self
+            except Exception as e:
+                self.log("Parameters are {0}".format(self.want))
+                self.msg = "Error in wireless provisioning due to {0}".format(e)
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:953]: Device ip/name

File: plugins/modules/provision_workflow_manager.py [hunk 7]
```diff
@@ -683,6 +785,47 @@ def get_want(self):
...
+            provision_info = self.get_execution_status_wireless(execution_id=execution_id)
+            self.result["changed"] = True
+            self.result['msg'] = "Wireless device got re-provisioned successfully"
+            self.result['diff'] = self.validated_config
+            self.result['response'] = execution_id
+            self.log(self.result['msg'], "INFO")
+            return self
+        except Exception as e:
+            self.log("Parameters are {0}".format(self.want))
+            self.msg = "Error in wireless re-provisioning due to {0}".format(e)
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:824]: device ip/name

File: plugins/modules/provision_workflow_manager.py (hunk 1 above)
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py:516]: Yes we have a bapi Error like this

File: plugins/modules/provision_workflow_manager.py (hunk 2 above)
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py]: Added

File: plugins/modules/provision_workflow_manager.py (hunk 4 above)
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py]: Addressed

File: plugins/modules/provision_workflow_manager.py (hunk 5 above)
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py]: Addressed

File: plugins/modules/provision_workflow_manager.py (hunk 6 above)
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py]: Addressed

File: plugins/modules/provision_workflow_manager.py (hunk 7 above)
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py]: Addressed

File: plugins/modules/provision_workflow_manager.py [hunk 8]
```diff
@@ -309,6 +333,34 @@ def get_dev_type(self):
...
+        """
+
+        dev_response = self.dnac_apply['exec'](
+            family="devices",
+            function='get_network_device_by_ip',
+            params={"ip_address": self.validated_config[0]["management_ip_address"]},
+            op_modifies=True
+        )
+
+        self.log("The device response from 'get_network_device_by_ip' API is {0}".format(str(dev_response)), "DEBUG")
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:357]: Can we print device IP address here?

File: plugins/modules/provision_workflow_manager.py [hunk 9]
```diff
@@ -309,6 +333,34 @@ def get_dev_type(self):
...
+        dev_dict = dev_response.get("response")
+        device_id = dev_dict.get("id")
+
+        self.log("Device ID of the device is {0}".format(device_id), "INFO")
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:361]: For which device?

File: plugins/modules/provision_workflow_manager.py [hunk 10]
```diff
@@ -789,13 +936,28 @@ class instance for further use.
...
+                provision_info = self.get_execution_status_wireless(execution_id=execution_id)
+                self.result["changed"] = True
+                self.result['msg'] = "Wireless device with IP {0} got provisioned successfully".format(self.validated_config[0]["management_ip_address"])
+                self.result['diff'] = self.validated_config
+                self.result['response'] = execution_id
+                self.log(self.result['msg'], "INFO")
+                return self
+            except Exception as e:
+                self.log("Parameters are {0}".format(self.want))
+                self.msg = "Error in wireless provisioning of {0}due to {1}".format(self.validated_config[0]["management_ip_address"], e)
```
Comment (madhansansel) [inline plugins/modules/provision_workflow_manager.py:957]: {0} can we leave a space after this..

File: plugins/modules/provision_workflow_manager.py (hunk 9 above)
Comment (abimishr) [reply plugins/modules/provision_workflow_manager.py]: Added

**Review by madhansansel — APPROVED** [review]

**Pull request merged**